<a href="https://colab.research.google.com/github/bdh777psu/UCSD-MLE-Bootcamp-Capstone/blob/main/Capstone_Step_5.ipynb" target="_parent"><img src="https://colab.research.google.com/assets/colab-badge.svg" alt="Open In Colab"/></a>

'Mental Health in the Tech Industry' is a complex and increasingly important issue as it encompasses many challenges with potentially severe impacts to one's well being: Fast-paced environments, high pressure, poor work-life balance, imposter syndrome, and lack of diversity can all play a significant role towards employee turnover, developing mental health issues, and even burnout.

Before starting to write any code and building the first model prototype for the Capstone project, what follows is an attempt to wrangle and clean published research data on the topic of 'Mental Health in the Tech Industry': The goal is to join the results of 'Open Sourcing Mental Health''s (OSMH) surveys on the topic into a single .csv file, containing data from years 2017-2023.


Step 1: Find out if the same (number of) questions were asked throughout 2017-2023

In [412]:
import pandas as pd

In [413]:
base_path = '/content/drive/MyDrive/UCSD-ML/Capstone/raw_data/'

In [414]:
df_2017 = pd.read_csv(base_path + 'responses_2017.csv')
df_2017.shape

(756, 123)

In [415]:
df_2018 = pd.read_csv(base_path + 'responses_2018.csv')
df_2018.shape

(417, 123)

In [416]:
df_2019 = pd.read_csv(base_path + 'responses_2019.csv')
df_2019.shape

(352, 82)

In [417]:
df_2020 = pd.read_csv(base_path + 'responses_2020.csv')
df_2020.shape

(180, 120)

In [418]:
df_2021 = pd.read_csv(base_path + 'responses_2021.csv')
df_2021.shape

(131, 124)

In [419]:
df_2022 = pd.read_csv(base_path + 'responses_2022.csv')
df_2022.shape

(164, 126)

In [420]:
df_2023 = pd.read_csv(base_path + 'responses_2023.csv')
df_2023.shape

(6, 126)

Findings: There was a significant change to the (number of) survey questions asked in the year 2019. Although, that year's responses could be dropped without too much impact to the final .csv data file, it contains 352 responses that can still be utilized with some data cleaning.

Step 2: Concatenating the results from survey years where the number of questions asked were relatively the same, starting with 2017-2018

In [421]:
import os

file_names = [base_path + 'responses_2017.csv', base_path + 'responses_2018.csv']
abs_paths = [os.path.join(file_name) for file_name in file_names]

cols = [pd.read_csv(abs_path, nrows=0).columns for abs_path in abs_paths]

cols_identical = [all(cols[0] == colx) for colx in cols[1:]]
all(cols_identical)


False

Findings: Roughly the same number of questions were asked in 2017-2018, however there are differences in the column headers, which need normalization before proceeding with concatenating both.

Step 3: Leave text-only data in the column headers, that comprise the actual questions that were asked in the respective surveys.

In [422]:
dirty_columns = {'<strong>Are you self-employed?</strong>': 'Are you self-employed?', '<strong>If you have been diagnosed or treated for a mental health disorder, do you ever reveal this to clients or business contacts?</strong>': 'If you have been diagnosed or treated for a mental health disorder, do you ever reveal this to clients or business contacts?', '<strong>If you have been diagnosed or treated for a mental health disorder, do you ever reveal this to coworkers or employees?</strong>': 'If you have been diagnosed or treated for a mental health disorder, do you ever reveal this to coworkers or employees?', '<strong>Do you have previous employers?</strong>': 'Do you have previous employers?', '<strong>Have your previous employers provided mental health benefits?</strong>': 'Have your previous employers provided mental health benefits?', '<strong>Were you aware of the options for mental health care provided by your previous employers?</strong>': 'Were you aware of the options for mental health care provided by your previous employers?', '<strong>Would you have been willing to discuss your mental health with your coworkers at previous employers?</strong>': 'Would you have been willing to discuss your mental health with your coworkers at previous employers?', 'If you have a mental health disorder, how often do you feel that it interferes with your work <strong>when being treated effectively?</strong>': 'If you have a mental health disorder, how often do you feel that it interferes with your work when being treated effectively?', '<strong>Have you observed or experienced an unsupportive or badly handled response to a mental health issue in your current or previous workplace?</strong>': 'Have you observed or experienced an unsupportive or badly handled response to a mental health issue in your current or previous workplace?', '<strong>Have you observed or experienced supportive or well handled response to a mental health issue in your current or previous workplace?</strong>': 'Have you observed or experienced supportive or well handled response to a mental health issue in your current or previous workplace?', 'What country do you <strong>live</strong> in?': 'What country do you live in?', 'What US state or territory do you <strong>live</strong> in?': 'What US state or territory do you live in?', 'What country do you <strong>work</strong> in?': 'What country do you work in?', 'What US state or territory do you <strong>work</strong> in?': 'What US state or territory do you work in?', 'Does your employer provide mental health benefits\xa0as part of healthcare coverage?': 'Does your employer provide mental health benefits as part of healthcare coverage?', 'If they knew you suffered from a mental health disorder,\xa0how do you think that team members/co-workers would react?': 'If they knew you suffered from a mental health disorder, how do you think that team members/co-workers would react?', 'If you have a mental health disorder, how often do you feel that it interferes with your work <strong>when <em>NOT</em> being treated effectively (i.e., when you are experiencing symptoms)?</strong>': 'If you have a mental health disorder, how often do you feel that it interferes with your work when NOT being treated effectively (i.e., when you are experiencing symptoms)?', 'If you have a mental health disorder, how often do you feel that it interferes with your work <strong>when</strong> <em><strong>NOT</strong></em><strong> being treated effectively (i.e., when you are experiencing symptoms)?</strong>': 'If you have a mental health disorder, how often do you feel that it interferes with your work when NOT being treated effectively (i.e., when you are experiencing symptoms)?', 'Would you be willing to talk to one of us more extensively about your experiences with mental health issues in the tech industry? (Note that all interview responses would be used <em>anonymously</em>\xa0and only with your permission.)': 'Would you be willing to talk to one of us more extensively about your experiences with mental health issues in the tech industry? (Note that all interview responses would be used anonymously and only with your permission.)'}

df_2017_clean = df_2017.rename(columns=dirty_columns)

In [423]:
df_2018_clean = df_2018.rename(columns=dirty_columns)

In [424]:
set(df_2017_clean.columns) - set(df_2018_clean.columns)

{'Would you be willing to talk to one of us more extensively about your experiences with mental health issues in the tech industry? (Note that all interview responses would be used anonymously and only with your permission.)'}

Findings: An empty set in the line above, indicates that there are no differences in the column headers from one year to the next (2017-2018)

Step 4: Concatenate the cleansed 2017 and 2018 survey results into a single dataframe, resulting in 1173 responses and 123 questions, so far.

In [425]:
data_frames = [df_2017_clean, df_2018_clean]

df_2017_2018 = pd.concat(data_frames)
df_2017_2018.shape

(1173, 124)

Step 5: Clean 2019 raw results and concatenate with 2017-2018 results.

(Extraneous open-ended questions that can be disregarded from surveys posterior to 2017-18 have been dropped from subsequent years, limiting dataset to 123 columns/questions at max.)

In [426]:
dirty_columns = {'*Are you self-employed?*': 'Are you self-employed?', '*Do you have previous employers?*': 'Do you have previous employers?', '*If possibly, what disorder(s) do you believe you have?*': 'If possibly, what disorder(s) do you believe you have?', '*If so, what disorder(s) were you diagnosed with?*': 'If so, what disorder(s) were you diagnosed with?', 'Do you *currently* have a mental health disorder?':  'Do you currently have a mental health disorder?', 'Have you ever been *diagnosed* with a mental health disorder?': 'Have you ever been diagnosed with a mental health disorder?', 'Have you ever been *diagnosed* with a mental health disorder?': 'Have you ever been diagnosed with a mental health disorder?', 'Have you observed or experienced a *supportive or well handled response* to a mental health issue in your current or previous workplace?': 'Have you observed or experienced a supportive or well handled response to a mental health issue in your current or previous workplace?', 'Have you observed or experienced an *unsupportive or badly handled response* to a mental health issue in your current or previous workplace?': 'Have you observed or experienced an *unsupportive or badly handled response to a mental health issue in your current or previous workplace?', 'Have you observed or experienced an *unsupportive or badly handled response* to a mental health issue in your current or previous workplace?': 'Have you observed or experienced an unsupportive or badly handled response to a mental health issue in your current or previous workplace?', 'If you have a mental health disorder, how often do you feel that it interferes with your work *when being treated effectively?*': 'If you have a mental health disorder, how often do you feel that it interferes with your work *when being treated effectively?', 'If you have a mental health disorder, how often do you feel that it interferes with your work *when* _*NOT*_* being treated effectively (i.e., when you are experiencing symptoms)?*': 'If you have a mental health disorder, how often do you feel that it interferes with your work when NOT being treated effectively (i.e., when you are experiencing symptoms)?', 'What US state or territory do you *live* in?': 'What US state or territory do you live in?', 'What US state or territory do you *work* in?': 'What US state or territory do you work in?', 'What country do you *live* in?': 'What country do you live in?', 'What country do you *work* in?': 'What country do you work in?', 'Would you be willing to talk to one of us more extensively about your experiences with mental health issues in the tech industry? (Note that all interview responses would be used _anonymously_ and only with your permission.)': 'Would you be willing to talk to one of us more extensively about your experiences with mental health issues in the tech industry? (Note that all interview responses would be used anonymously and only with your permission.)', 'Would you bring up your *mental* health with a potential employer in an interview?': 'Would you bring up your mental health with a potential employer in an interview?', '*What disorder(s) have you been diagnosed with?*': 'What disorder(s) have you been diagnosed with?', 'If you have a mental health disorder, how often do you feel that it interferes with your work *when being treated effectively?*': 'If you have a mental health disorder, how often do you feel that it interferes with your work when being treated effectively?'}

df_2019.rename(columns=dirty_columns, inplace=True)

In [427]:
extra_labels = set(df_2019) - set(df_2017_2018)

In [428]:
extra_labels_list = list(extra_labels)
df_2019_clean = df_2019.drop(columns=extra_labels_list)

In [429]:
set(df_2019_clean) - set(df_2017_2018)

set()

In [430]:
data_frames = [df_2017_2018, df_2019_clean]

df_2017_2019 = pd.concat(data_frames)
df_2017_2019.shape

(1525, 124)

Step 6: Clean 2020 raw results and concatenate with 2017-2019 results.

(Extraneous open-ended questions that can be disregarded from surveys posterior to 2017-18 have been dropped from subsequent years, constraining dataset to 123 columns/questions at max.)

In [431]:
df_2020.rename(columns=dirty_columns, inplace=True)

In [432]:
extra_labels = set(df_2020) - set(df_2017_2019)

In [433]:
extra_labels_list = list(extra_labels)
df_2020_clean = df_2020.drop(columns=extra_labels_list)

In [434]:
set(df_2020_clean) - set(df_2017_2019)

set()

In [435]:
data_frames = [df_2017_2019, df_2020_clean]

df_2017_2020 = pd.concat(data_frames)
df_2017_2020.shape

(1705, 124)

Step 7: Clean 2021 raw results and concatenate with 2017-2020 results.

(Extraneous open-ended questions that can be disregarded from surveys posterior to 2017-18 have been dropped from subsequent years, constraining dataset to 123 columns/questions at max.)

In [436]:
df_2021.rename(columns=dirty_columns, inplace=True)

In [437]:
extra_labels = set(df_2021) - set(df_2017_2020)

In [438]:
extra_labels_list = list(extra_labels)
df_2021_clean = df_2021.drop(columns=extra_labels_list)

In [439]:
set(df_2021_clean) - set(df_2017_2020)

set()

In [440]:
data_frames = [df_2017_2020, df_2021_clean]

df_2017_2021 = pd.concat(data_frames)
df_2017_2021.shape

(1836, 124)

Step 8: Clean 2022 raw results and concatenate with 2017-2021 results.

(Extraneous open-ended questions that can be disregarded from surveys posterior to 2017-18 have been dropped from subsequent years, constraining dataset to 123 columns/questions at max.)

In [441]:
df_2022.rename(columns=dirty_columns, inplace=True)

In [442]:
extra_labels = set(df_2022) - set(df_2017_2021)

In [443]:
extra_labels_list = list(extra_labels)
df_2022_clean = df_2022.drop(columns=extra_labels_list)

In [444]:
set(df_2021_clean) - set(df_2017_2020)

set()

In [445]:
data_frames = [df_2017_2021, df_2022_clean]

df_2017_2022 = pd.concat(data_frames)
df_2017_2022.shape

(2000, 124)

Step 9: Clean 2023 raw results and concatenate with 2017-2022 results.

(Extraneous open-ended questions that can be disregarded from surveys posterior to 2017-18 have been dropped from subsequent years, constraining dataset to 123 columns/questions at max.)

In [446]:
df_2023.rename(columns=dirty_columns, inplace=True)

In [447]:
extra_labels = set(df_2023) - set(df_2017_2022)

In [448]:
extra_labels_list = list(extra_labels)
df_2023_clean = df_2023.drop(columns=extra_labels_list)

In [449]:
set(df_2023_clean) - set(df_2017_2022)

set()

In [450]:
data_frames = [df_2017_2022, df_2023_clean]

df_2017_2023 = pd.concat(data_frames)
df_2017_2023.shape

(2006, 124)

Findings: There are now 2006 responses left from 123 total questions, however some are open ended questions which are not interesting for a quantitative analysis of the dataset an will be dropped.

Step 10: Drop columns containing open/ended survey questions and/or responses.

In [451]:
open_ended = ['Describe the conversation you had with your employer about your mental health, including their reactions and what actions were taken to address your mental health issue/questions.', 'Describe the conversation with coworkers you had about your mental health including their reactions.', 'Describe the conversation your coworker had with you about their mental health (please do not use names).', 'Describe the conversation you had with your previous employer about your mental health, including their reactions and actions taken to address your mental health issue/questions.', 'Describe the conversation you had with your previous coworkers about your mental health including their reactions.', 'Describe the conversation your coworker had with you about their mental health (please do not use names)..1', 'Why or why not?', 'Why or why not?.1', 'Describe the circumstances of the badly handled or unsupportive response.', 'Describe the circumstances of the supportive or well handled response.', 'Briefly describe what you think the industry as a whole and/or employers could do to improve mental health support for employees.', 'If there is anything else you would like to tell us that has not been covered by the survey questions, please use this space to do so.', 'Would you be willing to talk to one of us more extensively about your experiences with mental health issues in the tech industry? (Note that all interview responses would be used anonymously and only with your permission.)']
df_2017_2023.drop(columns=open_ended, inplace=True)

df_2017_2023.shape

(2006, 111)